sound/soc/sof/ipc4-topology.c

```diff
@@ -1357,18 +1358,24 @@ static int snd_sof_get_nhlt_endpoint_data(struct snd_sof_dev *sdev, struct snd_s
 						   &bit_depth);
 		if (ret < 0)
 			return ret;
+
+		/* for SSP, dev type could be 0 (BT sideband) or 4 (analog codec) */
+		dev_type = intel_nhlt_ssp_device_type(sdev->dev, ipc4_data->nhlt,
+						      dai_index);
+		if (ret < 0)
+			return ret;
 		break;
 	default:
 		return 0;
 	}
 
-	dev_dbg(sdev->dev, "dai index %d nhlt type %d direction %d\n",
-		dai_index, nhlt_type, dir);
+	dev_dbg(sdev->dev, "dai index %d nhlt type %d direction %d dev type %d\n",
+		dai_index, nhlt_type, dir, dev_type);
 
 	/* find NHLT blob with matching params */
 	cfg = intel_nhlt_get_endpoint_blob(sdev->dev, ipc4_data->nhlt, dai_index, nhlt_type,
 					   bit_depth, bit_depth, channel_count, sample_rate,
-					   dir, 0);
+					   dir, dev_type);
```

Comment:
Hrm so, this change basically makes the dev_type check kind of bypassed at the end?

What is expected if we have two sets of blobs for the same SSP index but one with dev_type=0 and the other with dev_type=4 in the NHLT blob? We will only going to be able to use the dev_type which is the first in the NHLT.

Is this acceptable and OK?

Reply:
Right, so this could be done in nhlt side just to ignore the dev_type?

I don't see how this dev_type have anything to do with the SSP configuration. SSP does not car if there is _anything_ connected to the I2S pins when it is the clock provider, in what way can possibly a mono, 16bit, 44.1KHz differs if the dev_type is BT offload or a codec?
Different FIFO thresholds probably?

Do we have examples of SSP NHLT blobs from BIOS? Does that contains different sets of configurations for the same SSP port per dev_type?

Reply:
After a bit of reading I think we are having this wrong when requesting the blob.
The endpoint description describes the _endpoint_ in a given system since I2S bus is not a discoverable bus.
So, you scan the SSP descriptors and you will have these metadata info about the connection, but for us what matters is the LinkType (SSP) Virtual Bus ID (the SSP port), the format that we are looking for and perhaps the direction.
The found descriptor will give additional info (if anyone cares) like the device_type, VendorID, DeviceID, etc.

The dev_type is not a selection parameter but it is a metadata in the NHLT which can be used by some OS, Linux is not using it.

@plbossart, @ranj063, @kv2019i, you know better the NHLT and things around it, do you see it differently?

Reply:
The difference device types were used by another OS to load different drivers. Indeed that's not something we use.

The only time we used the device type was to find which SSP was connected to the ES8336 codec. That wasn't to manage the SSP differently but rather select the relevant topology.

The point is that we need to ignore the device type and that's precisely what this PR does.